pkg/collector/corechecks/servicediscovery/impl_linux.go

```diff
@@ -0,0 +1,312 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux
+
+package servicediscovery
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/servicediscovery/portlist"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+//go:generate mockgen -source=$GOFILE -package=$GOPACKAGE -destination=impl_linux_mock.go
+
+func init() {
+	newOSImpl = newLinuxImpl
+}
+
+var ignoreCfgLinux = []string{
+	"sshd",
+	"dhclient",
+}
+
+type linuxImpl struct {
+	procfs     procFS
+	portPoller portPoller
+	time       timer
+
+	serviceDetector *serviceDetector
+	sender          *telemetrySender
+	ignoreCfg       map[string]bool
+
+	ignoreProcs       map[int]bool
+	aliveServices     map[int]*serviceInfo
+	potentialServices map[int]*serviceInfo
+}
+
+func newLinuxImpl(sender *telemetrySender, ignoreCfg map[string]bool) (osImpl, error) {
+	for _, i := range ignoreCfgLinux {
+		ignoreCfg[i] = true
+	}
+	pfs, err := procfs.NewDefaultFS()
+	if err != nil {
+		return nil, err
+	}
+	poller, err := portlist.NewPoller()
+	if err != nil {
+		return nil, err
+	}
+	return &linuxImpl{
+		procfs:            wProcFS{pfs},
+		portPoller:        poller,
+		time:              realTime{},
+		sender:            sender,
+		serviceDetector:   newServiceDetector(),
+		ignoreCfg:         ignoreCfg,
+		ignoreProcs:       make(map[int]bool),
+		aliveServices:     make(map[int]*serviceInfo),
+		potentialServices: make(map[int]*serviceInfo),
+	}, nil
+}
+
+type processEvents struct {
+	started []*serviceInfo
+	stopped []*serviceInfo
+}
+
+type eventsByName map[string]*processEvents
+
+func (e eventsByName) getOrCreate(name string) *processEvents {
+	events, ok := e[name]
+	if ok {
+		return events
+	}
+	e[name] = &processEvents{}
+	return e[name]
+}
+
+func (e eventsByName) addStarted(svc *serviceInfo) {
+	events := e.getOrCreate(svc.meta.Name)
+	events.started = append(events.started, svc)
```

Comment:
this is very confusing code. At first glance, a variable is created, a field in it is assigned a value, and then the variable is dead code. It works because events is a pointer to a value that was already assigned in e. 

The code is clearer if `getOrCreate` is rolled into the `addStarted` and `addStopped` functions:

```go
events, ok := e[svc.meta.Name]
if !ok {
    events = processEvents{}
}
events.started = append(events.started,svc)
e[svc.meta.Name] = events
```

I find this much easier to follow and the reduction in pointers makes less work for the garbage collector.